### PR TITLE
Add config error handler

### LINKED ISSUE #1667: Autocomplete: Show an error when an invalid model is configured
I have disabled auto-complete while working on something for a bit. But after that I re-enabled it but it stopped working. I have tried to enable/re-enable a few times but no luck. No auto-complete requests is being sent from the editor.

https://github.com/sourcegraph/cody/assets/44617923/7248d20e-e18b-4be9-a471-947187cd417c

My VSCode `settings.json`:

```JSON

{
  "workbench.colorTheme": "Sourcegraph Dark",
  "editor.unicodeHighlight.invisibleCharacters": false,
  "editor.unicodeHighlight.ambiguousCharacters": false,
  "cmake.configureOnOpen": true,
  "python.linting.pylintArgs": ["--max-line-length=130"],
  "[python]": {
    "editor.formatOnType": true
  },
  "redhat.telemetry.enabled": true,
  "editor.minimap.enabled": false,
  "go.toolsManagement.autoUpdate": true,
  "cody.debug.verbose": true,
  "cody.debug.enable": true,
  "hediet.vscode-drawio.resizeImages": null,
  "cody.autocomplete.advanced.cache": false,
  "cody.experimental.inline": true,
  "cody.inlineChat.enabled": true,
  "cody.autocomplete.experimental.triggerMoreEagerly": true,
  "cody.autocomplete.enabled": true,
  "editor.inlineSuggest.showToolbar": "always",
  "editor.inlineSuggest.suppressSuggestions": true,
  "cody.experimental.customRecipes": true,
  "[json]": {
    "editor.defaultFormatter": "vscode.json-language-features"
  },
  "cody.plugins.debug.enabled": false,
  "cody.experimental.editorTitleCommandIcon": true,
  "cody.experimental.commandLenses": true,
  "window.zoomLevel": 1,
  "typescript.updateImportsOnFileMove.enabled": "always",
  "cody.autocomplete.advanced.model": "claude-instant-infill",
  "cody.customHeaders": {},
  "git.autofetch": true,
  "diffEditor.ignoreTrimWhitespace": false,
  "[typescriptreact]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[css]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "workbench.tree.indent": 16,
  "cody.autocomplete.languages": {
    
    "*": true,
    "scminput": false
  }
}

```

CC:  @philipp-spiess Can you please help me with this?

### PR DESCRIPTION
**Pull request by deepak2431**
This PR closes #1667. 
I have implemented a function that checks for the allowed enum values for the Cody config. If the config value is not from the allowed list of enum values, then it shows an error during startup.

## Test plan

Open the `settings.json` and change `cody.autocomplete.advanced.provider` to `gpt-4`. After this, reload the window a VSCode error message will be displayed during the startup.

### CODE REVIEW

**Comment by deepak2431** [comment]
@philipp-spiess Can I have a review please?

File: vscode/src/configuration.ts [hunk 1]
```diff
@@ -54,9 +54,18 @@ export function getConfiguration(config: ConfigGetter = vscode.workspace.getConf
...
+    for (const configVal of configKeys) {
+        const key = configVal.replaceAll(/([A-Z])/g, '.$1').toLowerCase()
+        const value: string | null = config.get(CONFIG_KEY[configVal])
+        checkValidEnumValues(key, value)
+    }
+
     return {
-        // NOTE: serverEndpoint is now stored in Local Storage instead but we will still keep supporting the one in confg
-        // to use as fallback for users who do not have access to local storage
+        //  NOTE: serverEndpoint is now stored in Local Storage instead but we will still keep supporting the one in confg
```
Comment (philipp-spiess) [inline vscode/src/configuration.ts:67]: Can you revert the change that adds spaces after the comments? I think this was a bug (I also have it sometimes when I save files in VSCode, so weird!)

**Review by philipp-spiess — APPROVED** [review]
This is awesome thank you so much 🙇

**Comment by philipp-spiess** [comment]
CI is good: https://github.com/sourcegraph/cody/pull/1770

File: vscode/src/configuration.ts (hunk 1 above)
Comment (philipp-spiess) [inline vscode/src/configuration.ts:67]:
```suggestion
        // NOTE: serverEndpoint is now stored in Local Storage instead but we will still keep supporting the one in confg
```

File: vscode/src/configuration.ts [hunk 2]
```diff
@@ -54,9 +54,18 @@ export function getConfiguration(config: ConfigGetter = vscode.workspace.getConf
...
+        //  to use as fallback for users who do not have access to local storage
```
Comment (philipp-spiess) [inline vscode/src/configuration.ts:68]:
```suggestion
        // to use as fallback for users who do not have access to local storage
```

File: vscode/src/configuration.ts [hunk 3]
```diff
@@ -103,10 +112,10 @@ export function getConfiguration(config: ConfigGetter = vscode.workspace.getConf
          * UNDOCUMENTED FLAGS
          */
 
-        // Note: In spirit, we try to minimize agent-specific code paths in the VSC extension.
-        // We currently use this flag for the agent to provide more helpful error messages
-        // when something goes wrong, and to suppress event logging in the agent.
-        // Rely on this flag sparingly.
+        //  Note: In spirit, we try to minimize agent-specific code paths in the VSC extension.
```
Comment (philipp-spiess) [inline vscode/src/configuration.ts:115]:
```suggestion
        // Note: In spirit, we try to minimize agent-specific code paths in the VSC extension.
```

File: vscode/src/configuration.ts [hunk 4]
```diff
@@ -103,10 +112,10 @@ export function getConfiguration(config: ConfigGetter = vscode.workspace.getConf
...
+        //  We currently use this flag for the agent to provide more helpful error messages
```
Comment (philipp-spiess) [inline vscode/src/configuration.ts:116]:
```suggestion
        // We currently use this flag for the agent to provide more helpful error messages
```

File: vscode/src/configuration.ts [hunk 5]
```diff
@@ -103,10 +112,10 @@ export function getConfiguration(config: ConfigGetter = vscode.workspace.getConf
...
+        //  when something goes wrong, and to suppress event logging in the agent.
```
Comment (philipp-spiess) [inline vscode/src/configuration.ts:117]:
```suggestion
        // when something goes wrong, and to suppress event logging in the agent.
```

File: vscode/src/configuration.ts [hunk 6]
```diff
@@ -103,10 +112,10 @@ export function getConfiguration(config: ConfigGetter = vscode.workspace.getConf
...
+        //  Rely on this flag sparingly.
```
Comment (philipp-spiess) [inline vscode/src/configuration.ts:118]:
```suggestion
        // Rely on this flag sparingly.
```

File: vscode/src/configuration.ts [hunk 7]
```diff
@@ -123,8 +132,8 @@ function sanitizeCodebase(codebase: string | undefined): string {
 
 function sanitizeServerEndpoint(serverEndpoint: string): string {
     if (!serverEndpoint) {
-        // TODO(philipp-spiess): Find out why the config is not loaded properly in the integration
-        // tests.
+        //  TODO(philipp-spiess): Find out why the config is not loaded properly in the integration
```
Comment (philipp-spiess) [inline vscode/src/configuration.ts:135]:
```suggestion
        // TODO(philipp-spiess): Find out why the config is not loaded properly in the integration
```

File: vscode/src/configuration.ts [hunk 8]
```diff
@@ -123,8 +132,8 @@ function sanitizeCodebase(codebase: string | undefined): string {
...
+        //  tests.
```
Comment (philipp-spiess) [inline vscode/src/configuration.ts:136]:
```suggestion
        // tests.
```

File: vscode/src/configuration.ts [hunk 9]
```diff
@@ -138,8 +147,19 @@ function sanitizeServerEndpoint(serverEndpoint: string): string {
 
 export const getFullConfig = async (): Promise<ConfigurationWithAccessToken> => {
     const config = getConfiguration()
-    // Migrate endpoints to local storage
+    //  Migrate endpoints to local storage
```
Comment (philipp-spiess) [inline vscode/src/configuration.ts:150]:
```suggestion
    // Migrate endpoints to local storage
```

File: vscode/src/configuration.ts (hunk 1 above)
Comment (philipp-spiess) [reply vscode/src/configuration.ts]: (I was so free and pushed fixes for it so we can get it into this weeks release 😊 )

**Comment by deepak2431** [comment]
Thanks a lot, @philipp-spiess for the review and for making the comments space changes :)
